src/storage/store/engine.rs

```diff
@@ -813,6 +831,118 @@ impl ShardEngine {
+        };
+
+        // Determine user data type based on protocol
+        let user_data_type = match protocol {
+            proto::Protocol::Ethereum => UserDataType::UserDataPrimaryAddressEthereum,
+            proto::Protocol::Solana => UserDataType::UserDataPrimaryAddressSolana,
+        };
+
+        // Check if this address is set as the primary address
+        let user_data_result = UserDataStore::get_user_data_by_fid_and_type(
+            &self.stores.user_data_store,
+            fid,
+            user_data_type,
+        );
+
+        // If we can't find the user data or there's an error, just return OK (nothing to revoke)
+        let user_data = match user_data_result {
+            Ok(data) => data,
+            Err(_) => return Ok(()),
+        };
+
+        // Check if the stored primary address matches the one being removed
+        let data_ref = match &user_data.data {
+            Some(data) => data,
+            None => return Ok(()),
+        };
+
+        let user_data_body = match &data_ref.body {
+            Some(Body::UserDataBody(body)) => body,
+            _ => return Ok(()),
+        };
+
+        // If the primary address matches the one being removed, revoke it
+        if user_data_body.value == parsed_address {
+            self.stores.user_data_store.revoke(&user_data, txn_batch)?;
+        }
+
+        Ok(())
+    }
+
+    /// Takes a list of merge events of type VerificationAddAddress and revokes the user's primary address
+    /// if it was deleted as part of a verification remove message.
+    fn revoke_primary_addresses_for_deleted_verifications(
+        &mut self,
+        version: &EngineVersion,
+        events: &[HubEvent],
+        txn_batch: &mut RocksDbTransactionBatch,
+        source: &ProposalSource,
+    ) -> Result<(), EngineError> {
+        if !version.is_enabled(ProtocolFeature::PrimaryAddresses) {
+            return Ok(());
+        }
+
+        // Process verification removal hooks
+        // Look for any events with deleted verification messages. We use this as a trigger to revoke
+        // the user's primary address if it was deleted as part of a verification remove message.
+        for event in events {
+            if let Some(hub_event::Body::MergeMessageBody(merge_body)) = &event.body {
+                for deleted_message in &merge_body.deleted_messages {
+                    // Check if this was a verification add message that was deleted
+                    if let Some(data) = &deleted_message.data {
+                        // NOTE: VerificationAddEthAddress is used for both Ethereum and Solana verifications.
+                        // we differentiate between the two by checking the protocol field.
+                        if data.r#type == MessageType::VerificationAddEthAddress as i32 {
+                            if let Some(Body::VerificationAddAddressBody(body)) = &data.body {
```

Comment:
I think pattern matching is more concise here. Something like
```rust
       let data = deleted_message.data.as_ref().ok_or(MessageValidationError::NoMessageData)?
       match &data.body {
            Some(proto::message_data::Body::VerificationAddAddressBody(user_data)) => {
```

Also makes it easier to add more branches in the future